libraries/abstractions/common_io/test/test_iot_uart.c

```diff
@@ -471,30 +481,25 @@ TEST( TEST_IOT_UART, AFQP_IotUARTWriteReadAsyncWithCallback )
     int32_t lRead, lWrite, lClose;
     BaseType_t xCallbackReturn;
     uint8_t ucPort = uctestIotUartPort;
-    volatile IotUARTOperationStatus_t xCallbackStatus = eUartLastWriteFailed;
 
     xUartHandle = iot_uart_open( ucPort );
     TEST_ASSERT_NOT_EQUAL( NULL, xUartHandle );
 
     if( TEST_PROTECT() )
     {
-        iot_uart_set_callback( xUartHandle, prvReadWriteCallback, ( void * ) &xCallbackStatus );
+        iot_uart_set_callback( xUartHandle, prvReadWriteCallback, NULL );
 
         lWrite = iot_uart_write_async( xUartHandle, cpBuffer, testIotUART_BUFFER_LENGTH );
         TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lWrite );
 
-        xCallbackReturn = xSemaphoreTake( ( SemaphoreHandle_t ) &xtestIotUARTCompleted, testIotUART_DEFAULT_SEMPAHORE_DELAY );
-        TEST_ASSERT_EQUAL( pdTRUE, xCallbackReturn );
-
-        TEST_ASSERT_EQUAL( eUartWriteCompleted, xCallbackStatus );
-
         lRead = iot_uart_read_async( xUartHandle, cpBufferRead, testIotUART_BUFFER_LENGTH );
         TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lRead );
 
-        xCallbackReturn = xSemaphoreTake( ( SemaphoreHandle_t ) &xtestIotUARTCompleted, testIotUART_DEFAULT_SEMPAHORE_DELAY );
+        xCallbackReturn = xSemaphoreTake( xWriteCompleteSemaphore, testIotUART_DEFAULT_SEMPAHORE_DELAY );
```

Comment:
Should we wait on the write semaphore before scheduling the read? How do we expect a real customer app to do this?

Reply:
No. That is the exact point I am changing in this PR. This test case is designed to be a wire-loopback one so the read should be scheduled immediately after write. Otherwise, data comes without being read. That caused UART overrun error.

For real customer app, it is unlikely to do loopback. That case is test by assisted tests (e.g. with Raspberry Pi) that not covered in this PR.